[ffmpeg] add CMake targets to FindFFMPEG.cmake module

Be-ing opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  
 missing CMake targets for ffmpeg. I ran into issues using `${FFMPEG_LIBRARIES}` in a generator expression like `$<$<BOOL:${USE_FFMPEG}>:${FFMPEG_LIBRARIES}>` because `${FFMPEG_LIBRARIES}` is a list. Using a target resolves that and also makes it easier to add the include directories.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  all

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  I am still working on this PR and asking for help.

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

At the end of the build, this is printed:
```
The package ffmpeg:x64-linux provides CMake targets:

    find_package(ffmpeg CONFIG REQUIRED)
    target_link_libraries(main PRIVATE FFMPEG::FFMPEG)
```

I am getting an error I don't understand when doing `find_package(FFMPEG CONFIG REQUIRED)` in the application's CMakeLists.txt:
```
CMake Error at /home/be/sw/vcpkg/installed/x64-linux/share/ffmpeg/vcpkg-cmake-wrapper.cmake:6 (_find_package):
  Could not find a package configuration file provided by "FFMPEG" with any
  of the following names:

    FFMPEGConfig.cmake
    ffmpeg-config.cmake

  Add the installation prefix of "FFMPEG" to CMAKE_PREFIX_PATH or set
  "FFMPEG_DIR" to a directory containing one of the above files.  If "FFMPEG"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  /home/be/sw/vcpkg/scripts/buildsystems/vcpkg.cmake:736 (include)
  CMakeLists.txt:559 (find_package)


-- Configuring incomplete, errors occurred!
```

If I change it to `find_package(ffmpeg CONFIG REQURIED)` that does not work either:
```
CMake Error at /home/be/sw/vcpkg/installed/x64-linux/share/ffmpeg/vcpkg-cmake-wrapper.cmake:6 (_find_package):
  Could not find a package configuration file provided by "ffmpeg" with any
  of the following names:

    ffmpegConfig.cmake
    ffmpeg-config.cmake

  Add the installation prefix of "ffmpeg" to CMAKE_PREFIX_PATH or set
  "ffmpeg_DIR" to a directory containing one of the above files.  If "ffmpeg"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  /home/be/sw/vcpkg/scripts/buildsystems/vcpkg.cmake:736 (include)
  CMakeLists.txt:559 (find_package)


-- Configuring incomplete, errors occurred!
```

Be-ing committed:
[ffmpeg] add CMake targets to FindFFMPEG.cmake module

Be-ing commented:
CI failures seem to be unrelated?

Be-ing commented:
Working on another port, it seems vcpkg has issues if the target exported by a CMake config file from a library's build system does not exactly match the port's name, including the case I guess.

mcmtroffaes commented:
I think it's a great idea to provide a target.

> I am getting an error I don't understand when doing find_package(FFMPEG CONFIG REQUIRED) in the application's CMakeLists.txt:

There's a simple explanation for this. [CMake has two methods for finding packages](https://cmake.org/cmake/help/latest/command/find_package.html): via FindXXX.cmake, or via XXXConfig.cmake. Currently FFmpeg uses the FindXXX method for historical reasons, although it probably should switch to a XXXConfig.cmake system (at least in my opinion). Using the "CONFIG" option forces cmake to look only for the XXXConfig.cmake script which obviously fails since it does not exist.

Not related to the suggested changes but related to the topic of the PR, note that, if you want to get targets, you can also import ffmpeg using [pkgconfig](https://cmake.org/cmake/help/latest/module/FindPkgConfig.html). This has the added flexibility that you can directly specify which components of ffmpeg you want to link to (and even their individual versions). For example:

```
find_package(PkgConfig REQUIRED)
pkg_check_modules(AVFORMAT REQUIRED IMPORTED_TARGET libavformat libswresample)
add_executable(main main.cpp)
target_link_libraries(main PRIVATE PkgConfig::AVFORMAT)
```

For this to work on windows, you need to manually point cmake to recognize pkgconf on windows; for instance:

```
vcpkg install pkgconf:x64-windows
set PKG_CONFIG=c:\vcpkg\installed\x64-windows\tools\pkgconf\pkgconf.exe
```

mcmtroffaes commented on ports/ffmpeg/FindFFMPEG.cmake.in:
```diff
@@ -89,6 +89,17 @@ macro(FFMPEG_FIND varname shortname headername)
+    if(NOT TARGET FFMPEG::FFMPEG)
+      add_library(FFMPEG::FFMPEG INTERFACE IMPORTED)
+    endif()
+    target_link_libraries(FFMPEG::FFMPEG PUBLIC FFMPEG::${varname})
```
If I understand correctly, this code only adds ``FFMPEG_${varname}_LIBRARY`` to each target. What about the dependencies?

Be-ing replied:
Using a target with `target_link_libraries` includes all the transitive dependencies.

dg0yt replied:
But it is not magic. It is based on explicit information. In your PR, you add a `FFMPEG::${varname}` target, and almost all what is known about it is `INTERFACE_LINK_LIBRARIES FFMPEG_${varname}_LIBRARY`. Even if you would properly dereference the variable name (hint!), it would not cover a single transitive dependency (unless the variable contains another target which is fully configured).

dg0yt reviewed (commented):
If you add targets which are neither used by upstream nor by genuine CMake find modules, they should use the namespace `unofficial`.

Be-ing commented:
> Currently FFmpeg uses the FindXXX method for historical reasons, although it probably should switch to a XXXConfig.cmake system (at least in my opinion). Using the "CONFIG" option forces cmake to look only for the XXXConfig.cmake script which obviously fails since it does not exist.

Ah, that makes sense. I don't think I have seen FindXXXX.cmake modules in other vcpkg ports.

The motivation that lead to opening this PR is that I was having trouble linking FFmpeg in a generator expression:

```
$<$<BOOL:${USE_FFMPEG}>:${FFMPEG_LIBRARIES}>
```

That gave some strange errors that I don't recall. I think they were because `${FFMPEG_LIBRARIES}` is a list. I couldn't figure out how to get that to work in a generator expression. So I modified the FindFFMPEG.cmake module in the application to create targets and changed it to:

```
$<$<BOOL:${USE_FFMPEG}>:FFMPEG::FFMPEG>
```

then it worked.

mcmtroffaes commented:
> That gave some strange errors that I don't recall. I think they were because ${FFMPEG_LIBRARIES} is a list.

Sounds like a syntax issue. Have you tried putting quotes around the expression, e.g. ``"$<$<BOOL:${USE_FFMPEG}>:${FFMPEG_LIBRARIES}>"``? I'm not sure if that will fix the problem, but might be worth a shot.

Be-ing commented:
This would be nice to have but I'm not particularly motivated to continue working on this myself, so I'm closing this. I would appreciate if someone else picked this up though.
